test/celeritas/optical/ValidationUtils.hh

```diff
@@ -0,0 +1,186 @@
+ */
+template<class UnitType, class... Args>
+std::array<real_type, sizeof...(Args)> constexpr expressed_as(Args const&... args)
+{
+    return std::array<real_type, sizeof...(Args)>{
+        native_value_from(UnitType(args))...};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function to annotate units of hard-coded test data.
+ *
+ * Same as \c expressed_as, except returns a vector.
+ */
+template<class UnitType, class... Args>
+std::vector<real_type> vector_expressed_as(Args const&... args)
+{
+    return std::vector<real_type>{native_value_from(UnitType(args))...};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Checks two \c ImportPhysicVector for exact equality.
+ */
+void check_physics_vector(ImportPhysicsVector const& expected,
+                          ImportPhysicsVector const& actual);
+
+//---------------------------------------------------------------------------//
+/*!
+ * Convert a floating point type collection into a \c real_type span.
+ *
+ * Used to convert data which may be expressed as double or single precision
+ * into the same size as \c real_type. If the supplied data is the same
+ * precision, then just a span to the original data is returned. Otherwise,
+ * it is copied into a local buffer and casted to the correct precision. The
+ * \c RealSpanGenerator object should live as long as its spans are being
+ * used in the tests.
+ */
+class RealSpanGenerator
+{
+  public:
+    /*!
+     * Create a \c real_type \c Span for the supplied collection type \c T.
+     *
+     * If \c T::value_type is \c real_type, then a span to the original data
+     * is returned. Otherwise, it is copied to a local buffer and cast to
+     * \c real_type.
+     */
+    template<class T>
+    Span<real_type const> operator()(T const& xs)
+    {
+        if constexpr (std::is_same_v<real_type,
+                                     std::remove_cv_t<typename T::value_type>>)
+        {
+            return make_span(xs);
+        }
+        else
+        {
+            buffer_.resize(xs.size());
+            for (auto i : range(xs.size()))
+            {
+                buffer_[i] = static_cast<real_type>(xs[i]);
+            }
+            return make_span(buffer_);
+        }
+    }
+
+  private:
+    std::vector<real_type> buffer_;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Perform consistency checks on grids built in \c Collections.
+ */
+class GridValidator
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Grid = GenericGridRecord;
+    using GridId = OpaqueId<Grid>;
+    using ImportPhysicsTable = std::vector<ImportPhysicsVector>;
+
+    template<class T>
+    using Items = Collection<T, Ownership::value, MemSpace::host>;
+    //!@}
+
+    //! Whether to use soft or exact equivalence
+    enum Softness
+    {
+        Soft,
+        Exact
+    };
+
```

Comment:
I definitely don't like this. If anything, make it a templated function with a comparator that could use SoftEqual or std::equal .

Reply:
I ended up not using soft equality, so I've removed this enum.